ops/terraform/AOU_RW_MODULE_WALKTHROUGH.md

```diff
@@ -0,0 +1,277 @@
+Terraform will collect all `.tf` files in a referenced directory,
+so the calling module will need to specify values for the chilid
+modules' `variable` blocks that don't have defaults.
+
+```text
+/Users/jaycarlton/repos/workbench/ops/terraform/
+├── AOU_RW_MODULE_WALKTHROUGH.md
+├── TERRAFORM-QUICKSTART.md
+├── environments
+│   ├── local
+│   ├── scratch
+│   │   ├── SCRATCH-ENVIRONMENT.md
+│   │   ├── scratch.tf
+│   │   ├── terraform.tfstate
+│   │   ├── terraform.tfstate.backup
+│   │   └── terraform.tfstate.yet.another.backup
+│   └── test
+└── modules
+    └── aou-rw-reporting
+        ├── providers.tf
+        ├── reporting.tf
+        ├── schemas
+        │   ├── cohort.json
+        │   ├── institution.json
+        │   ├── user.json
+        │   └── workspace.json
+        ├── variables.tf
+        └── views
+            ├── latest_cohorts.sql
+            ├── latest_institutions.sql
+            ├── latest_users.sql
+            ├── latest_workspaces.sql
+            └── table_count_vs_time.sql
+```
+The `modules` directory contains independent, reusable modules foor
+subsystems that are 
+* logical to deploy and configure operationally,
+* don't depend on each other (at least for exported modules) and
+* can be used by AoU or potentially another organization interested in deploying a copy
+of all or part of our system.
+
+## Prerequisites
+### 1. Get Terraform
+Install Terraform using the directinos at [TERRAFORM-QUICKSTART.md]
+### 2. Change to the `environments/scratch directory` `get` and `init`
+The environment for this outline is `scratch`, which exists in a target environment
+of your choice. 
+### 3. Assign Values to Input Variables
+
+The following public variable declarations are representative of those
+specified in `modules/reporting/variables.tf` and elsewhere. The description
+string shows when interactively running from the command line without all the
+vars cominig in from a `-var-file` argument.
+```hcl-terraform
+variable credentials_file {
+  description = "Location of service account credentials JSON file."
+  type        = string
+}
+
+variable aou_env {
+  description = "Short name (all lowercase) of All of Us Workbench deployed environments, e.g. local, test, staging, prod."
+  type        = string
+}
+
+variable project_id {
+  description = "GCP Project"
+  type        = string
+}
+```
+Create a `scratch_tutorrial.tfvars` file outside of this repository. This file should
+look contain values for the following [input variables](https://www.terraform.io/docs/configuration/variables.html) that will be different
+for different organizations and environments.
+
+```hcl-terraform
+aou_env = "scratch" # Name of environment we're creating or attaching to. Needs to match directory name
+project_id = "my-qa-project" # Should not be prod
+reporting_dataset_id = "firstname_lastname_scratch_0" # BigQuery  dataset id
+credentials_file = "~/scratch/keys/ops-sa-key.json"
```

Comment:
Better to use either ADC, or this SA token approach: https://github.com/hashicorp/terraform-provider-google/issues/4642#issuecomment-540101168 ; keyfiles should be avoided where possible, and it's a lot less work to setup/teardown

Reply:
So for context, does this work now?
```
provider "google" {
  impersonate_service_account: "foo@bar.iam.gserviceaccount.com"
}
```

I meant to put a comment here asking you for the preferred approach. I won't miss key files.

Reply:
No, it's the name of an authentication method/library for gcloud which makes auth more consistent in different environments (local, GAE, GCE, etc). You would use [this command](https://cloud.google.com/sdk/gcloud/reference/auth/application-default/login) locally with your pmi-ops.org; this apparently is the default bheavior if you just omit the "credentials_file" entirely to the Google TF provider.

The other method I mentioned involves more boilerplate. The syntax I proposed on that issue was not implemented - instead you can use the approach demonstrated on the example code on this page: https://www.terraform.io/docs/providers/google/d/service_account_access_token.html ; this would give you a temporary SA credential - but probably you'd need to use in combination with ADC in order to get the token to start.

Reply:
Can we leave the auth for a follow-on PR? This work isn't in the critical path yet (i.e. `bq` and the console is enough for reporting's current needs), but I am looking to aggressively leverage Terraform for Multi-tier stuff.